crates/relayer/src/keyring/ed25519_key_pair.rs

```diff
@@ -42,50 +42,82 @@ fn standard_path_to_derivation_path(path: &StandardHDPath) -> DerivationPath {
     DerivationPath::new(child_indices)
 }
 
+#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
+pub enum Ed25519AddressType {
+    Solana,
+}
+
+impl TryFrom<&AddressType> for Ed25519AddressType {
+    type Error = Error;
+
+    fn try_from(address_type: &AddressType) -> Result<Self, Self::Error> {
+        match address_type {
+            AddressType::Cosmos | AddressType::Ethermint { .. } => Err(
+                Error::unsupported_address_type(address_type.clone(), Ed25519KeyPair::KEY_TYPE),
+            ),
+        }
+    }
+}
+
 #[derive(Debug, Deserialize, Serialize)]
 pub struct Ed25519KeyPair {
     keypair: Keypair,
+    address_type: Ed25519AddressType,
 }
 
 impl Clone for Ed25519KeyPair {
     fn clone(&self) -> Self {
         let Self {
             keypair: Keypair { secret, public },
+            address_type,
         } = self;
         Self {
             keypair: Keypair {
+                // Needed as `SecretKey` does not implement `Clone`
                 secret: SecretKey::from_bytes(secret.as_bytes()).unwrap(),
                 public: *public,
             },
+            address_type: *address_type,
         }
     }
 }
 
 impl Ed25519KeyPair {
-    fn from_mnemonic_internal(mnemonic: &str, hd_path: &StandardHDPath) -> Result<Self, Error> {
+    fn from_mnemonic_internal(
+        mnemonic: &str,
+        hd_path: &StandardHDPath,
+        address_type: Ed25519AddressType,
+    ) -> Result<Self, Error> {
         let private_key = private_key_from_mnemonic(mnemonic, hd_path)?.secret_key;
         let public_key = (&private_key).into();
         let keypair = Keypair {
             secret: private_key,
             public: public_key,
         };
 
-        Ok(Self { keypair })
+        Ok(Self {
+            keypair,
+            address_type,
+        })
     }
 }
 
 impl SigningKeyPair for Ed25519KeyPair {
     const KEY_TYPE: KeyType = KeyType::Ed25519;
 
     fn from_key_file(key_file: KeyFile, hd_path: &StandardHDPath) -> Result<Self, Error> {
-        let key_pair = Self::from_mnemonic_internal(&key_file.mnemonic, hd_path)?;
-
-        let public_key_from_file = PublicKey::from_bytes(
-            &bs58::decode(key_file.pubkey)
-                .into_vec()
-                .map_err(Error::bs58_decode)?,
-        )
-        .map_err(Error::invalid_public_key)?;
+        // TODO: Derive this from something in `key_file`
```

Comment:
What do you think about adding this to `KeyPairV2`, so that we don't have to introduce a `KeyPairV3` for this change in the future?

Reply:
Hmm this is in the Ed25519 side (Solana) and I think I've already added this to the `KeyPair` type so the only thing that would need to be changed here is updating what we're reading from `KeyFile`. So we should be able to avoid adding a `KeyPairV2` here.